monai/data/meta_obj.py

```diff
@@ -0,0 +1,172 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from copy import deepcopy
+from typing import Any, Callable, Sequence
+
+import numpy as np
+import torch
+
+_TRACK_META = True
+_TRACK_TRANSFORMS = True
+
+__all__ = ["get_track_meta", "get_track_transforms", "set_track_meta", "set_track_transforms", "MetaObj"]
+
+
+def set_track_meta(val: bool) -> None:
+    """
+    Boolean to set whether metadata is tracked. If `True`,
+    `MetaTensor` will be returned where appropriate. If `False`,
+    `torch.Tensor` will be returned instead.
+    """
+    global _TRACK_META
+    _TRACK_META = val
+
+
+def set_track_transforms(val: bool) -> None:
+    """
+    Boolean to set whether transforms are tracked.
+    """
+    global _TRACK_TRANSFORMS
+    _TRACK_TRANSFORMS = val
+
+
+def get_track_meta() -> bool:
+    """
+    Get track meta data boolean.
+    """
+    global _TRACK_META
+    return _TRACK_META
+
+
+def get_track_transforms() -> bool:
+    """
+    Get track transform boolean.
+    """
+    global _TRACK_TRANSFORMS
+    return _TRACK_TRANSFORMS
+
+
+class MetaObj:
+    """
+    Class that stores meta and affine.
+
+    We store the affine as its own element, so that this can be updated by
+    transforms. All other meta data that we don't plan on touching until we
+    need to save the image to file lives in `meta`.
+
+    This allows for subclassing `np.ndarray` and `torch.Tensor`.
+
+    Copying metadata:
+        * For `c = a + b`, then the meta data will be copied from the first
+        instance of `MetaImage`.
+    """
+
+    _meta: dict
+    _affine: torch.Tensor
+
+    def set_initial_val(self, attribute: str, input_arg: Any, input_tensor: MetaObj, default_fn: Callable) -> None:
+        """
+        Set the initial value. Try to use input argument, but if this is None
+        and there is a MetaImage input, then copy that. Failing both these two,
+        use a default value.
+        """
+        if input_arg is None:
+            input_arg = getattr(input_tensor, attribute, None)
+        if input_arg is None:
+            input_arg = default_fn(self)
+        setattr(self, attribute, input_arg)
+
+    @staticmethod
+    def get_tensors_or_arrays(args: Sequence[Any]) -> list[MetaObj]:
```

Comment:
I feel this function is more like "flatten data"? And no need to mention `tensor` or `array` in this base class.

Thanks.